pkg/controller/volume/reconciler/reconciler.go

```diff
@@ -75,44 +76,56 @@ func (rc *reconciler) Run(stopCh <-chan struct{}) {
-				_, err := rc.actualStateOfWorld.AddVolumeNode(
-					volumeToAttach.VolumeSpec, volumeToAttach.NodeName)
-				if err != nil {
-					glog.Errorf("Unexpected error on actualStateOfWorld.AddVolumeNode(): %v", err)
-				}
-			} else {
-				// Volume/Node doesn't exist, spawn a goroutine to attach it
-				glog.V(5).Infof("Triggering AttachVolume for volume %q to node %q", volumeToAttach.VolumeName, volumeToAttach.NodeName)
-				rc.attacherDetacher.AttachVolume(&volumeToAttach, rc.actualStateOfWorld)
-			}
-		}
+		// Detaches are triggered before attaches so that volumes referenced by
+		// pods that are rescheduled to a different node are detached first.
 
 		// Ensure volumes that should be detached are detached.
 		for _, attachedVolume := range rc.actualStateOfWorld.GetAttachedVolumes() {
 			if !rc.desiredStateOfWorld.VolumeExists(
 				attachedVolume.VolumeName, attachedVolume.NodeName) {
 				// Volume exists in actual state of world but not desired
-				if attachedVolume.SafeToDetach {
-					glog.V(5).Infof("Triggering DetachVolume for volume %q to node %q", attachedVolume.VolumeName, attachedVolume.NodeName)
-					rc.attacherDetacher.DetachVolume(&attachedVolume, rc.actualStateOfWorld)
+				if !attachedVolume.MountedByNode {
+					glog.V(5).Infof("Attempting to start DetachVolume for volume %q to node %q", attachedVolume.VolumeName, attachedVolume.NodeName)
+					err := rc.attacherDetacher.DetachVolume(attachedVolume, rc.actualStateOfWorld)
+					if err == nil {
+						glog.Infof("Started DetachVolume for volume %q to node %q", attachedVolume.VolumeName, attachedVolume.NodeName)
+					}
 				} else {
-					// If volume is not safe to detach wait a max amount of time before detaching any way.
+					// If volume is not safe to detach (is mounted) wait a max amount of time before detaching any way.
 					timeElapsed, err := rc.actualStateOfWorld.MarkDesireToDetach(attachedVolume.VolumeName, attachedVolume.NodeName)
 					if err != nil {
 						glog.Errorf("Unexpected error actualStateOfWorld.MarkDesireToDetach(): %v", err)
 					}
-					if timeElapsed > rc.maxSafeToDetachDuration {
-						glog.V(5).Infof("Triggering DetachVolume for volume %q to node %q. Volume is not safe to detach, but max wait time expired.", attachedVolume.VolumeName, attachedVolume.NodeName)
-						rc.attacherDetacher.DetachVolume(&attachedVolume, rc.actualStateOfWorld)
+					if timeElapsed > rc.maxWaitForUnmountDuration {
+						glog.V(5).Infof("Attempting to start DetachVolume for volume %q to node %q. Volume is not safe to detach, but maxWaitForUnmountDuration expired.", attachedVolume.VolumeName, attachedVolume.NodeName)
+						err := rc.attacherDetacher.DetachVolume(attachedVolume, rc.actualStateOfWorld)
+						if err == nil {
+							glog.Infof("Started DetachVolume for volume %q to node %q. Volume was not safe to detach, but maxWaitForUnmountDuration expired.", attachedVolume.VolumeName, attachedVolume.NodeName)
```

Comment:
Suggestion: rephrase this as `started DetachVolume for ... due to maxWaitForUnmountDuration expiry`